executor/mem_reader.go

```diff
@@ -143,12 +143,15 @@ func (m *memIndexReader) getMemRows(ctx context.Context) ([][]types.Datum, error
 	}
 
 	if m.keepOrder && m.table.GetPartitionInfo() != nil {
-		slices.SortFunc(m.addedRows, func(a, b []types.Datum) bool {
+		slices.SortFunc(m.addedRows, func(a, b []types.Datum) int {
 			ret, err1 := m.compare(m.ctx.GetSessionVars().StmtCtx, a, b)
```

Comment:
Better change the `compare` function

https://github.com/pingcap/tidb/blob/4fc7970f216e6ff0e2c9f09d07a9545ba128243c/executor/union_scan.go#L295